### 1) Calcule as integrais a seguir pela regra dos trapézios e 1/3 de Simpson, usando quatro e seis divisões de $[a, b]$. Obtenha um limitante superior para o erro cometido e compare com o valor exato, quando possível.

#### (a) $\int_1^2 e^x \, dx$

$$
\int_1^2 e^x \, dx = \frac{b-a}{2n} \left( f(a) + f(b) + 2 \sum_{i=1}^{n-1} f(x_i) \right) \\
\int_1^2 e^x \, dx = \frac{2-1}{2(4)} \left( f(1) + f(2) + 2 \sum_{i=1}^{4-1} f(x_i) \right) \\
\int_1^2 e^x \, dx = \frac{1}{8} \left( e + e^2 + 2 \sum_{i=1}^{3} f(x_i) \right)
$$

In [1]:
from numpy import linspace
def regra_trapezio(f, a, b, intervalos):
    soma = f(a) + f(b) # inicializa a soma com o valor de f(a) + f(b)
    
    for x in linspace(a, b, intervalos+1)[1:-1]: # divide [a, b] em intervalos
        soma += 2 * f(x)
    
    delta_x = (b-a)/intervalos
    
    return (delta_x/2) * soma

In [2]:
from numpy import linspace
def simpson_terco(f, a, b, intervalos):
    delta_x = (b-a)/intervalos # tamanho dos intervalos

    soma = f(a) + f(b)
    for i, x in enumerate(linspace(a, b, intervalos+1)[1:-1], start=1):
        if i % 2 == 0:
            soma += 2 * f(x)
        else:
            soma += 4 * f(x)

    return (delta_x/3) * soma

In [3]:
from math import exp
print(f'Trapézio com 4 intervalos:        {regra_trapezio(exp, 1, 2, 4)}')
print(f'Trapézio com 6 intervalos:        {regra_trapezio(exp, 1, 2, 6)}')
print(f'1/3 de Simpson com 4 intervalos:  {simpson_terco(exp, 1, 2, 4)}')
print(f'1/3 de Simpson com 6 intervalos:  {simpson_terco(exp, 1, 2, 6)}')

Trapézio com 4 intervalos:        4.695075916875121
Trapézio com 6 intervalos:        4.6815812457126675
1/3 de Simpson com 4 intervalos:  4.670874883494676
1/3 de Simpson com 6 intervalos:  4.670794226633774


Agora calculando o erro da Regra do Trapézio
$$
R_T(n) \leq \frac{M (b - a)^3}{12n^2}
$$
Onde $M = \mathrm{max} |f''(x)|$ para $x \in [a, b]$ e $n$ é o número de intervalos.

$$
f''(x) = e^x
$$

Caso 1: $a = 1$, $b = 2$ e $n = 4$

$$
M = e^2 \\
R_T(4) \leq \frac{e^2 (2-1)^3}{12 \cdot 4^2} \\
R_T(4) \leq \frac{e^2}{192}
$$

In [4]:
exp(2) / 192

0.03848466718193047

$$
R_T(4) \leq 0.038485
$$

Caso 2: $a = 1$, $b = 2$ e $n = 6$

$$
R_T(6) \leq \frac{e^2 (2-1)^3}{12 \cdot 6^2} \\
$$

In [5]:
exp(2) / (12 * 6**2)

0.01710429652530243

$$
R_T(6) \leq 0.017104
$$

O erro da Regra de Simpson é dado por
$$
R_S(n) = \frac{M (b-a)^5}{180n^4}
$$
Onde $M = \mathrm{max} |f^{(4)}(x)|$ para $x \in [a, b]$ e $n$ é o número de intervalos.

Caso 1: $a = 1$, $b = 2$ e $n = 4$

$$
M = e^2 \\
R_S(4) \leq \frac{e^2 (2-1)^5}{180 \cdot 4^4} \\
$$

In [6]:
exp(2) / (180 * 4**4)

0.0001603527799247103

$$
R_S(4) \leq 0.000160
$$

Caso 2: $a = 1$, $b = 2$ e $n = 6$

$$
R_S(6) \leq \frac{e^2 (2-1)^5}{180 \cdot 6^4} \\
$$

In [7]:
exp(2) / (180 * 6**4)

3.16746231950045e-05

$$
R_S(6) \leq 3.16 \cdot 10^{-5}
$$

#### (b) $\int_1^4 \sqrt{x} \, dx$

In [8]:
from math import sqrt
print(f'Trapézio com 4 intervalos:        {regra_trapezio(sqrt, 1, 4, 4)}')
print(f'Trapézio com 6 intervalos:        {regra_trapezio(sqrt, 1, 4, 6)}')
print(f'1/3 de Simpson com 4 intervalos:  {simpson_terco(sqrt, 1, 4, 4)}')
print(f'1/3 de Simpson com 6 intervalos:  {simpson_terco(sqrt, 1, 4, 6)}')

Trapézio com 4 intervalos:        4.65509259251136
Trapézio com 6 intervalos:        4.661488382402361
1/3 de Simpson com 4 intervalos:  4.666220708306385
1/3 de Simpson com 6 intervalos:  4.6665630532224895


In [9]:
def erro_trapezio(a, b, M, n):
    return (M * (b-a)**3) / (12 * n**2)

In [10]:
def erro_simpson(a, b, M, n):
    return (M * (b-a)**5) / (180 * n**4)

Encontrando os erros, começando com os Trapézios:

$$
M = \mathrm{max} |f''(x)|, x \in [1, 4] \\
M = \mathrm{max} \left| \frac{-1}{4x^{3/2}} \right|, x \in [1, 4] \\
$$

O ponto que maximiza a função é $x = 1$.

$$
M = \frac{1}{4}
$$

Calculando os erros

In [11]:
print(erro_trapezio(1, 4, 1/4, 4))
print(erro_trapezio(1, 4, 1/4, 6))

0.03515625
0.015625


Então
$$
R_T(4) \leq 0.035156 \\
R_T(6) \leq 0.015625
$$

Agora os erros de Simpson

$$
M = \mathrm{max} |f^{(4)}(x)|, x \in [1, 4] \\
M = \mathrm{max} \left| \frac{-15}{16 x^{7/2}} \right|, x \in [1, 4]
$$

Novamente, $x = 1$ maximiza a função.

$$
M = \frac{15}{16}
$$

Calculando os erros

In [12]:
print(erro_simpson(1, 4, 15/16, 4))
print(erro_simpson(1, 4, 15/16, 6))

0.00494384765625
0.0009765625


Portanto
$$
R_S(4) \leq 0.004944 \\
R_S(6) \leq 0.000977
$$

#### (c) $\int_2^{14} \frac{1}{\sqrt{x}} \, dx$

In [13]:
from math import sqrt
f = lambda x: 1/sqrt(x)
print(f'Trapézio com 4 intervalos:        {regra_trapezio(f, 2, 14, 4)}')
print(f'Trapézio com 6 intervalos:        {regra_trapezio(f, 2, 14, 6)}')
print(f'1/3 de Simpson com 4 intervalos:  {simpson_terco(f, 2, 14, 4)}')
print(f'1/3 de Simpson com 4 intervalos:  {simpson_terco(f, 2, 14, 6)}')

Trapézio com 4 intervalos:        4.768387026661444
Trapézio com 6 intervalos:        4.707777186436547
1/3 de Simpson com 4 intervalos:  4.676374564596406
1/3 de Simpson com 4 intervalos:  4.66148949120848


Encontrando os erros, começando pelos Trapézios

$$
M = \mathrm{max} |f''(x)|, x \in [2, 14] \\
M = \mathrm{max} \left| \frac{3}{4 x^{5/2}} \right|, x \in [2, 14] \\
$$

Essa função é maximizada em $x = 2$.

$$
M = \frac{3}{16 \sqrt{2}}
$$

Calculando os erros

In [14]:
print(erro_trapezio(2, 14, 3/(16*sqrt(2)), 4))
print(erro_trapezio(2, 14, 3/(16*sqrt(2)), 6))

1.1932426932522988
0.5303300858899106


Portanto
$$
R_T(4) \leq 1.193243 \\
R_T(6) \leq 0.530330
$$

Agora os erros de Simpson

$$
M = \mathrm{max} |f^{(4)}(x)|, x \in [1, 4] \\
M = \mathrm{max} \left| \frac{105}{16 x^{9/2}} \right|, x \in [1, 4]
$$

Novamente, o máximo ocorre em $x = 2$.

$$
M = \frac{105}{256 \sqrt{2}}
$$

Calculando os erros

In [15]:
print(erro_simpson(1, 4, 105/(256*sqrt(2)), 4))
print(erro_simpson(1, 4, 105/(256*sqrt(2)), 6))

0.0015294248387633222
0.00030210861012608837


Portanto
$$
R_S(4) \leq 0.001529 \\
R_S(6) \leq 0.000302
$$

### 4) Dada a tabela:

In [16]:
x  = [0.0,    0.2,    0.4,    0.6,    0.8,    1.0]
fx = [1.0, 1.2408, 1.5735, 2.0333, 2.6965, 3.7183]

e sabendo que a regra 1/3 de Simpson é, em geral, mais precisa que a dos trapézios, qual seria o modo mais adequado de calcular $I = \int_0^1 f(x) \, dx$, usando a tabela acima? Aplique este processo e determine esta integral.

Como o número de intervalos é ímpar, não se pode usar a regra de Simpson diretamente. Para contornar isso, pode-se usar a regra do trapézio no primeiro intervalo e somar com a regra de Simpson nos intervalos restantes.

Começando pela regra do trapézio
$$
\frac{h}{2} (f(x_0) + f(x_1)) \\
\frac{0.2}{2} (1 + 1.2408) = 0.22408
$$

E o restante será

In [17]:
def simpson_terco_discreto(x, fx):
    delta_x = x[1] - x[0] # descobre o tamanho do intervalo
    
    soma = fx[0] + fx[-1]
    for i, y in enumerate(fx[1:-1], start=1):
        if i % 2 == 0:
            soma += 2 * y
        else:
            soma += 4 * y
    
    return (delta_x/3) * soma

In [18]:
simpson_terco_discreto(x[1:], fx[1:])

1.7403799999999998

Então
$$
I = 1.74038 + 0.22408 \\
I = 1.96446
$$